Crash upon trying to start second download while download in progress

CKAN v1.16.1-0-q2e91715, OS X

If CKAN is in the middle of a manual mod download, going back to the "Manage mods" tab and trying to manually download another mod causes a crash

To recreate: Open up CKAN, click on the EVE config file mod, go to contents -> Download. While download is progressing, return to "Manage mods" tab, find FAR, and go to contents -> Download.

Output of stdout/stderr:

```
982 [1] ERROR CKAN.URLHandlers (null) - There was an error while registering the URL handler for ckan:// - Could not find a part of the path "/Users/awang/.local/share/applications/mimeapps.list".
993 [1] ERROR CKAN.URLHandlers (null) -   at System.IO.FileStream..ctor (System.String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, Boolean anonymous, FileOptions options) <0x1ffbaa0 + 0x0021d> in <filename unknown>:0 
  at System.IO.FileStream..ctor (System.String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, System.String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost) <0x1ffb9d0 + 0x0004f> in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) System.IO.FileStream:.ctor (string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare,int,System.IO.FileOptions,string,bool,bool,bool)
  at System.IO.StreamWriter.CreateFile (System.String path, Boolean append, Boolean checkHost) <0x1e50c30 + 0x000a4> in <filename unknown>:0 
  at System.IO.StreamWriter..ctor (System.String path, Boolean append, System.Text.Encoding encoding, Int32 bufferSize, Boolean checkHost) <0x1e508b0 + 0x0006c> in <filename unknown>:0 
  at System.IO.StreamWriter..ctor (System.String path, Boolean append, System.Text.Encoding encoding, Int32 bufferSize) <0x1e50860 + 0x00042> in <filename unknown>:0 
  at System.IO.StreamWriter..ctor (System.String path) <0x1e50770 + 0x0003d> in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) System.IO.StreamWriter:.ctor (string)
  at System.IO.File.WriteAllLines (System.String path, System.String[] contents) <0x1ff9b30 + 0x0003b> in <filename unknown>:0 
  at CKAN.URLHandlers.RegisterURLHandler_Linux () <0x82f24e8 + 0x00137> in <filename unknown>:0 
  at CKAN.URLHandlers.RegisterURLHandler (CKAN.Configuration config, IUser user) <0x82f2208 + 0x00027> in <filename unknown>:0 
System.InvalidOperationException: This BackgroundWorker is currently busy and cannot run multiple tasks concurrently.
  at System.ComponentModel.BackgroundWorker.RunWorkerAsync (System.Object argument) <0x9abdb80 + 0x000af> in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) System.ComponentModel.BackgroundWorker:RunWorkerAsync (object)
  at CKAN.Main.ContentsDownloadButton_Click (System.Object sender, System.EventArgs e) <0xa994e50 + 0x0006f> in <filename unknown>:0 
  at System.Windows.Forms.Control.OnClick (System.EventArgs e) <0xa171ff8 + 0x00056> in <filename unknown>:0 
  at System.Windows.Forms.Button.OnClick (System.EventArgs e) <0xa994de0 + 0x0004b> in <filename unknown>:0 
  at System.Windows.Forms.ButtonBase.OnMouseUp (System.Windows.Forms.MouseEventArgs mevent) <0xa994c90 + 0x00123> in <filename unknown>:0 
  at System.Windows.Forms.Button.OnMouseUp (System.Windows.Forms.MouseEventArgs mevent) <0xa994c68 + 0x00017> in <filename unknown>:0 
  at System.Windows.Forms.Control.WmLButtonUp (System.Windows.Forms.Message& m) <0xa171dc0 + 0x00128> in <filename unknown>:0 
  at System.Windows.Forms.Control.WndProc (System.Windows.Forms.Message& m) <0x3d2dee0 + 0x001c3> in <filename unknown>:0 
  at System.Windows.Forms.ButtonBase.WndProc (System.Windows.Forms.Message& m) <0x9ac13c0 + 0x0005f> in <filename unknown>:0 
  at System.Windows.Forms.Button.WndProc (System.Windows.Forms.Message& m) <0x9ac1398 + 0x00017> in <filename unknown>:0 
  at System.Windows.Forms.Control+ControlWindowTarget.OnMessage (System.Windows.Forms.Message& m) <0x3d2dec0 + 0x0001d> in <filename unknown>:0 
  at System.Windows.Forms.Control+ControlNativeWindow.WndProc (System.Windows.Forms.Message& m) <0x3d2de90 + 0x0002b> in <filename unknown>:0 
  at System.Windows.Forms.NativeWindow.WndProc (IntPtr hWnd, Msg msg, IntPtr wParam, IntPtr lParam) <0x3d2ca90 + 0x002cb> in <filename unknown>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: Cannot call Invoke or BeginInvoke on a control until the window handle is created
  at System.Windows.Forms.Control.FindControlToInvokeOn () <0xa165c08 + 0x000c7> in <filename unknown>:0 
  at System.Windows.Forms.Control.Invoke (System.Delegate method, System.Object[] args) <0xa9953e0 + 0x00017> in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) System.Windows.Forms.Control:Invoke (System.Delegate,object[])
  at System.Windows.Forms.Control.Invoke (System.Delegate method) <0xa995278 + 0x000a7> in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) System.Windows.Forms.Control:Invoke (System.Delegate)
  at CKAN.Util.Invoke[T] (CKAN.T obj, System.Action action) <0x47ca510 + 0x0009b> in <filename unknown>:0 
  at CKAN.Main.SetDescription (System.String message) <0x9abe900 + 0x000fb> in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) CKAN.Main:SetDescription (string)
  at CKAN.GUIUser.ReportProgress (System.String format, Int32 percent) <0xa99ac08 + 0x000eb> in <filename unknown>:0 
  at CKAN.NullUser.RaiseProgress (System.String message, Int32 percent) <0xa99abe0 + 0x00022> in <filename unknown>:0 
  at CKAN.NetAsyncDownloader.FileProgressReport (Int32 index, Int32 percent, Int64 bytesDownloaded, Int64 bytesToDownload) <0xa99a0a8 + 0x00837> in <filename unknown>:0 
  at CKAN.NetAsyncDownloader+<DownloadCurl>c__AnonStorey1.<>m__0 (System.Object extraData, Double dlTotal, Double dlNow, Double ulTotal, Double ulNow) <0xa999cc0 + 0x001e7> in <filename unknown>:0 
  at CurlSharp.CurlEasy._curlProgressCallback (IntPtr ptrUserdata, Double dlTotal, Double dlNow, Double ulTotal, Double ulNow) <0xa999be0 + 0x0005b> in <filename unknown>:0 
  at (wrapper native-to-managed) CurlSharp.CurlEasy:_curlProgressCallback (intptr,double,double,double,double)
  at (wrapper managed-to-native) CurlSharp.NativeMethods:curl_easy_perform (intptr)
  at CurlSharp.CurlEasy.Perform () <0xa999b20 + 0x00027> in <filename unknown>:0 
  at CKAN.NetAsyncDownloader.CurlWatchThread (Int32 index, CurlSharp.CurlEasy easy, System.IO.FileStream stream) <0xa9999e8 + 0x0002f> in <filename unknown>:0 
  at CKAN.NetAsyncDownloader+<DownloadCurl>c__AnonStorey1.<>m__1 () <0xa9999a0 + 0x0002f> in <filename unknown>:0 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) <0x1f420e0 + 0x0008e> in <filename unknown>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, Boolean preserveSyncCtx) <0x1f407b0 + 0x001a4> in <filename unknown>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, Boolean preserveSyncCtx) <0x1f40770 + 0x00033> in <filename unknown>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) <0x1f406c0 + 0x0005a> in <filename unknown>:0 
  at System.Threading.ThreadHelper.ThreadStart () <0x1f42260 + 0x00039> in <filename unknown>:0 
exception inside UnhandledException handler: Cannot call Invoke or BeginInvoke on a control until the window handle is created
```

--- a/Core/Net/NetAsyncDownloader.cs
+++ b/Core/Net/NetAsyncDownloader.cs
@@ -8,7 +8,6 @@ using System.Text.RegularExpressions;
 using System.Threading;
 using Autofac;
 using CKAN.Configuration;
-using CurlSharp;
 using log4net;
 
 namespace CKAN
@@ -114,10 +113,6 @@ namespace CKAN
         public delegate void NetAsyncOneCompleted(Uri url, string filename, Exception error);
         public NetAsyncOneCompleted onOneCompleted;
 
-        // When using the curlsharp downloader, this contains all the threads
-        // that are working for us.
-        private List<Thread> curl_threads = new List<Thread>();
-
         /// <summary>
         /// Returns a perfectly boring NetAsyncDownloader.
         /// </summary>
@@ -138,174 +133,56 @@ namespace CKAN
             downloads.Clear();
             foreach (Net.DownloadTarget target in targets)
             {
-                downloads.Add(new NetAsyncDownloaderDownloadPart(target));
+                DownloadModule(target);
             }
+        }
 
-            // adding chicken bits
-            if (Platform.IsWindows || System.Environment.GetEnvironmentVariable("KSP_CKAN_USE_CURL") == null)
-            {
-                DownloadNative();
-            }
-            else
-            {
-                DownloadCurl();
-            }
+        private void DownloadModule(Net.DownloadTarget target)
+        {
+            // We need a new variable for our closure/lambda, hence index = 1+prev max
+            int index = downloads.Count;
 
+            var dl = new NetAsyncDownloaderDownloadPart(target);
+            downloads.Add(dl);
+
+            // Encode spaces to avoid confusing URL parsers
+            User.RaiseMessage("Downloading \"{0}\"",
+                dl.url.ToString().Replace(" ", "%20"));
+
+            // Schedule for us to get back progress reports.
+            dl.Progress += (sender, args) =>
+                FileProgressReport(index,
+                    args.ProgressPercentage,
+                    args.BytesReceived,
+                    args.TotalBytesToReceive);
+
+            // And schedule a notification if we're done (or if something goes wrong)
+            dl.Done += (sender, args) =>
+                FileDownloadComplete(index, args.Error);
+
+            // Start the download!
+            dl.Download(dl.url, dl.path);
         }
 
         /// <summary>
-        /// Download all our files using the native .NET handlers.
+        /// Start a new batch of downloads
         /// </summary>
-        /// <returns>The native.</returns>
-        private void DownloadNative()
-        {
-            for (int i = 0; i < downloads.Count; i++)
-            {
-                // Encode spaces to avoid confusing URL parsers
-                User.RaiseMessage("Downloading \"{0}\"",
-                    downloads[i].url.ToString().Replace(" ", "%20"));
-
-                // We need a new variable for our closure/lambda, hence index = i.
-                int index = i;
-
-                // Schedule for us to get back progress reports.
-                downloads[i].Progress += (sender, args) =>
-                    FileProgressReport(index,
-                        args.ProgressPercentage,
-                        args.BytesReceived,
-                        args.TotalBytesToReceive);
-
-                // And schedule a notification if we're done (or if something goes wrong)
-                downloads[i].Done += (sender, args) =>
-                    FileDownloadComplete(index, args.Error);
-
-                // Start the download!
-                downloads[i].Download(downloads[i].url, downloads[i].path);
-            }
-        }
-
-        /// <summary>
-        /// Use curlsharp to handle our downloads.
-        /// </summary>
-        private void DownloadCurl()
-        {
-            log.Debug("Curlsharp async downloader engaged");
-
-            // Make sure our environment is set up.
-
-            Curl.Init();
-
-            // We'd *like* to use CurlMulti, but it just hangs when I try to retrieve
-            // messages from it. So we're spawning a thread for each curleasy that does
-            // the same thing. Ends up this is a little easier in handling, anyway.
-
-            for (int i = 0; i < downloads.Count; i++)
-            {
-                log.DebugFormat("Downloading {0}", downloads[i].url);
-                // Encode spaces to avoid confusing URL parsers
-                User.RaiseMessage("Downloading \"{0}\" (libcurl)",
-                    downloads[i].url.ToString().Replace(" ", "%20"));
-
-                // Open our file, and make an easy object...
-                FileStream stream = File.OpenWrite(downloads[i].path);
-                CurlEasy easy = Curl.CreateEasy(downloads[i].url, stream);
-
-                // We need a separate variable for our closure, this is it.
-                int index = i;
-
-                // Curl recommends xferinfofunction, but this doesn't seem to
-                // be supported by curlsharp, so we use the progress function
-                // instead.
-                easy.ProgressFunction = delegate(object extraData, double dlTotal, double dlNow, double ulTotal, double ulNow)
-                {
-                    log.DebugFormat("Progress function called... {0}/{1}", dlNow,dlTotal);
-
-                    int percent;
-
-                    if (dlTotal > 0)
-                    {
-                        percent = (int) dlNow * 100 / (int) dlTotal;
-                    }
-                    else
-                    {
-                        log.Debug("Unknown download size, skipping progress.");
-                        return 0;
-                    }
-
-                    FileProgressReport(
-                        index,
-                        percent,
-                        Convert.ToInt64(dlNow),
-                        Convert.ToInt64(dlTotal)
-                    );
-
-                    // If the user has told us to cancel, then bail out now.
-                    if (download_canceled)
-                    {
-                        log.InfoFormat("Bailing out of download {0} at user request", index);
-                        // Bail out!
-                        return 1;
-                    }
-
-                    // Returning 0 means we want to continue the download.
-                    return 0;
-                };
-
-                // Download, little curl, fulfill your destiny!
-                Thread thread = new Thread(new ThreadStart(delegate
-                {
-                    CurlWatchThread(index, easy, stream);
-                }));
-
-                // Keep track of our threads so we can clean them up later.
-                curl_threads.Add(thread);
-
-                // Background threads will mostly look after themselves.
-                thread.IsBackground = true;
-
-                // Let's go!
-                thread.Start();
-            }
-        }
-
-        /// <summary>
-        /// Starts a thread to watch download progress. Invoked by DownloadCUrl. Not for
-        /// public consumption.
-        /// </summary>
-        private void CurlWatchThread(int index, CurlEasy easy, FileStream stream)
-        {
-            log.Debug("Curlsharp download thread started");
-
-            // This should run until completion or failture.
-            CurlCode result = easy.Perform();
-
-            log.Debug("Curlsharp download complete");
-
-            // Dispose of all our disposables.
-            // We have to do this *BEFORE* we call FileDownloadComplete, as it
-            // ensure we've written everything out to disk.
-            stream.Dispose();
-            easy.Dispose();
-
-            if (result == CurlCode.Ok)
-            {
-                FileDownloadComplete(index, null);
-            }
-            else
-            {
-                // The CurlCode result expands to a human-friendly string, so we can just
-                // throw a kraken containing it and nothing else. The FileDownloadComplete
-                // code collects these into a larger DownloadErrorsKraken aggregate.
-
-                FileDownloadComplete(
-                    index,
-                    new Kraken(result.ToString())
-                );
-            }
-        }
-
+        /// <param name="urls">The downloads to begin</param>
         public void DownloadAndWait(ICollection<Net.DownloadTarget> urls)
         {
+            if (downloads.Count > completed_downloads)
+            {
+                // Some downloads are still in progress, add to the current batch
+                foreach (Net.DownloadTarget target in urls)
+                {
+                    DownloadModule(target);
+                }
+                // Wait for completion along with original caller
+                // so we can handle completion tasks for the added mods
+                complete_or_canceled.WaitOne();
+                return;
+            }
+
             completed_downloads = 0;
             // Make sure we are ready to start a fresh batch
             complete_or_canceled.Reset();
@@ -326,20 +203,12 @@ namespace CKAN
 
 
             // If the user cancelled our progress, then signal that.
-            // This *should* be harmless if we're using the curlsharp downloader,
-            // which watches for downloadCanceled all by itself. :)
             if (old_download_canceled)
             {
                 // Abort all our traditional downloads, if there are any.
                 foreach (var download in downloads.ToList())
                 {
                     download.Abort();
-                }
-
-                // Abort all our curl downloads, if there are any.
-                foreach (var thread in curl_threads.ToList())
-                {
-                    thread.Abort();
                 }
 
                 // Signal to the caller that the user cancelled the download.
@@ -466,9 +335,8 @@ namespace CKAN
         }
 
         /// <summary>
-        /// This method gets called back by `WebClient` or our
-        /// curl downloader when a download is completed. It in turn
-        /// calls the onCompleted hook when *all* downloads are finished.
+        /// This method gets called back by `WebClient` when a download is completed.
+        /// It in turncalls the onCompleted hook when *all* downloads are finished.
         /// </summary>
         private void FileDownloadComplete(int index, Exception error)
         {

--- a/Core/Net/NetAsyncDownloader.cs
+++ b/Core/Net/NetAsyncDownloader.cs
@@ -124,10 +124,9 @@ namespace CKAN
         }
 
         /// <summary>
-        /// Downloads our files, returning an array of filenames that we're writing to.
-        /// The sole argument is a collection of DownloadTargets.
-        /// The .onCompleted delegate will be called on completion.
+        /// Downloads our files.
         /// </summary>
+        /// <param name="targets">A collection of DownloadTargets</param>
         private void Download(ICollection<Net.DownloadTarget> targets)
         {
             downloads.Clear();

--- a/Core/Net/NetAsyncModulesDownloader.cs
+++ b/Core/Net/NetAsyncModulesDownloader.cs
@@ -29,7 +29,11 @@ namespace CKAN
         public NetAsyncModulesDownloader(IUser user, NetModuleCache cache)
         {
             modules    = new List<CkanModule>();
-            downloader = new NetAsyncDownloader(user);
+            downloader = new NetAsyncDownloader(user)
+            {
+                // Schedule us to process each module on completion.
+                onOneCompleted = ModuleDownloadComplete
+            };
             this.cache = cache;
         }
 
@@ -39,16 +43,15 @@ namespace CKAN
         public void DownloadModules(IEnumerable<CkanModule> modules)
         {
             // Walk through all our modules, but only keep the first of each
-            // one that has a unique download path.
+            // one that has a unique download path (including active downloads).
+            var currentlyActive = new HashSet<Uri>(this.modules.Select(m => m.download));
             Dictionary<Uri, CkanModule> unique_downloads = modules
                 .GroupBy(module => module.download)
-                .ToDictionary(p => p.First().download, p => p.First());
+                .Where(group => !currentlyActive.Contains(group.Key))
+                .ToDictionary(group => group.Key, group => group.First());
 
             this.modules.AddRange(unique_downloads.Values);
 
-            // Schedule us to process each module on completion.
-            downloader.onOneCompleted = ModuleDownloadComplete;
-                    
             try
             {
                 // Start the downloads!

--- a/Core/Net/NetModuleCache.cs
+++ b/Core/Net/NetModuleCache.cs
@@ -162,6 +162,18 @@ namespace CKAN
             return cache.Store(module.download, path, description ?? module.StandardName(), move);
         }
 
+        /// <summary>
+        /// Remove a module's download file from the cache
+        /// </summary>
+        /// <param name="module">Module to purge</param>
+        /// <returns>
+        /// True if purged, false otherwise
+        /// </returns>
+        public bool Purge(CkanModule module)
+        {
+            return cache.Remove(module.download);
+        }
+
         private NetFileCache cache;
     }
 }

--- a/GUI/Localization/de-DE/Main.de-DE.resx
+++ b/GUI/Localization/de-DE/Main.de-DE.resx
@@ -189,7 +189,8 @@
   <data name="FilterInstalledButton.Text" xml:space="preserve"><value>Installiert</value></data>
   <data name="FilterInstalledUpdateButton.Text" xml:space="preserve"><value>Aktualisierbar</value></data>
   <data name="FilterReplaceableButton.Text" xml:space="preserve"><value>Ersetzbar</value></data>
-  <data name="cachedToolStripMenuItem.Text" xml:space="preserve"><value>Im Cache</value></data>
+  <data name="FilterCachedButton.Text" xml:space="preserve"><value>Im Cache</value></data>
+  <data name="FilterUncachedButton.Text" xml:space="preserve"><value>Nicht im Cache</value></data>
   <data name="FilterNewButton.Text" xml:space="preserve"><value>Neu kompatibel</value></data>
   <data name="FilterNotInstalledButton.Text" xml:space="preserve"><value>Nicht installiert</value></data>
   <data name="FilterIncompatibleButton.Text" xml:space="preserve"><value>Inkompatibel</value></data>
@@ -208,6 +209,7 @@
   <data name="Description.HeaderText" xml:space="preserve"><value>Kurzbeschreibung</value></data>
   <data name="reinstallToolStripMenuItem.Text" xml:space="preserve"><value>Neu installieren</value></data>
   <data name="downloadContentsToolStripMenuItem.Text" xml:space="preserve"><value>Inhalt herunterladen</value></data>
+  <data name="purgeContentsToolStripMenuItem.Text" xml:space="preserve"><value>Inhalt löschen</value></data>
   <data name="ManageModsTabPage.Text" xml:space="preserve"><value>Mods verwalten</value></data>
   <data name="FilterByAuthorLabel.Text" xml:space="preserve"><value>Nach Autorenname filtern:</value></data>
   <data name="FilterByAuthorTextBox.Location" type="System.Drawing.Point, System.Drawing"><value>570, 74</value></data>

--- a/GUI/Main.Designer.cs
+++ b/GUI/Main.Designer.cs
@@ -62,7 +62,8 @@
             this.FilterInstalledButton = new System.Windows.Forms.ToolStripMenuItem();
             this.FilterInstalledUpdateButton = new System.Windows.Forms.ToolStripMenuItem();
             this.FilterReplaceableButton = new System.Windows.Forms.ToolStripMenuItem();
-            this.cachedToolStripMenuItem = new System.Windows.Forms.ToolStripMenuItem();
+            this.FilterCachedButton = new System.Windows.Forms.ToolStripMenuItem();
+            this.FilterUncachedButton = new System.Windows.Forms.ToolStripMenuItem();
             this.FilterNewButton = new System.Windows.Forms.ToolStripMenuItem();
             this.FilterNotInstalledButton = new System.Windows.Forms.ToolStripMenuItem();
             this.FilterIncompatibleButton = new System.Windows.Forms.ToolStripMenuItem();
@@ -89,6 +90,7 @@
             this.ModListHeaderContextMenuStrip = new System.Windows.Forms.ContextMenuStrip(this.components);
             this.reinstallToolStripMenuItem = new System.Windows.Forms.ToolStripMenuItem();
             this.downloadContentsToolStripMenuItem = new System.Windows.Forms.ToolStripMenuItem();
+            this.purgeContentsToolStripMenuItem = new System.Windows.Forms.ToolStripMenuItem();
             this.ModInfoTabControl = new CKAN.MainModInfo();
             this.StatusLabel = new System.Windows.Forms.ToolStripStatusLabel();
             this.StatusInstanceLabel = new System.Windows.Forms.ToolStripStatusLabel();
@@ -416,7 +418,8 @@
             this.FilterInstalledButton,
             this.FilterInstalledUpdateButton,
             this.FilterReplaceableButton,
-            this.cachedToolStripMenuItem,
+            this.FilterCachedButton,
+            this.FilterUncachedButton,
             this.FilterNewButton,
             this.FilterNotInstalledButton,
             this.FilterIncompatibleButton,
@@ -455,12 +458,19 @@
             this.FilterReplaceableButton.Click += new System.EventHandler(this.FilterReplaceableButton_Click);
             resources.ApplyResources(this.FilterReplaceableButton, "FilterReplaceableButton");
             //
-            // cachedToolStripMenuItem
+            // FilterCachedButton
             //
-            this.cachedToolStripMenuItem.Name = "cachedToolStripMenuItem";
-            this.cachedToolStripMenuItem.Size = new System.Drawing.Size(307, 30);
-            this.cachedToolStripMenuItem.Click += new System.EventHandler(this.cachedToolStripMenuItem_Click);
-            resources.ApplyResources(this.cachedToolStripMenuItem, "cachedToolStripMenuItem");
+            this.FilterCachedButton.Name = "FilterCachedButton";
+            this.FilterCachedButton.Size = new System.Drawing.Size(307, 30);
+            this.FilterCachedButton.Click += new System.EventHandler(this.FilterCachedButton_Click);
+            resources.ApplyResources(this.FilterCachedButton, "FilterCachedButton");
+            //
+            // FilterUncachedButton
+            //
+            this.FilterUncachedButton.Name = "FilterUncachedButton";
+            this.FilterUncachedButton.Size = new System.Drawing.Size(307, 30);
+            this.FilterUncachedButton.Click += new System.EventHandler(this.FilterUncachedButton_Click);
+            resources.ApplyResources(this.FilterUncachedButton, "FilterUncachedButton");
             //
             // FilterNewButton
             //
@@ -689,7 +699,8 @@
             //
             this.ModListContextMenuStrip.Items.AddRange(new System.Windows.Forms.ToolStripItem[] {
             this.reinstallToolStripMenuItem,
-            this.downloadContentsToolStripMenuItem});
+            this.downloadContentsToolStripMenuItem,
+            this.purgeContentsToolStripMenuItem});
             this.ModListContextMenuStrip.Name = "ModListContextMenuStrip";
             this.ModListContextMenuStrip.Size = new System.Drawing.Size(180, 70);
             //
@@ -706,6 +717,13 @@
             this.downloadContentsToolStripMenuItem.Size = new System.Drawing.Size(179, 22);
             this.downloadContentsToolStripMenuItem.Click += new System.EventHandler(this.downloadContentsToolStripMenuItem_Click);
             resources.ApplyResources(this.downloadContentsToolStripMenuItem, "downloadContentsToolStripMenuItem");
+            //
+            // purgeContentsToolStripMenuItem
+            //
+            this.purgeContentsToolStripMenuItem.Name = "purgeContentsToolStripMenuItem";
+            this.purgeContentsToolStripMenuItem.Size = new System.Drawing.Size(179, 22);
+            this.purgeContentsToolStripMenuItem.Click += new System.EventHandler(this.purgeContentsToolStripMenuItem_Click);
+            resources.ApplyResources(this.purgeContentsToolStripMenuItem, "purgeContentsToolStripMenuItem");
             //
             // ModListHeaderContextMenuStrip
             //
@@ -1401,7 +1419,8 @@
         private System.Windows.Forms.ToolStripMenuItem FilterInstalledButton;
         private System.Windows.Forms.ToolStripMenuItem FilterInstalledUpdateButton;
         private System.Windows.Forms.ToolStripMenuItem FilterReplaceableButton;
-        private System.Windows.Forms.ToolStripMenuItem cachedToolStripMenuItem;
+        private System.Windows.Forms.ToolStripMenuItem FilterCachedButton;
+        private System.Windows.Forms.ToolStripMenuItem FilterUncachedButton;
         private System.Windows.Forms.ToolStripMenuItem FilterNewButton;
         private System.Windows.Forms.ToolStripMenuItem FilterNotInstalledButton;
         private System.Windows.Forms.ToolStripMenuItem FilterIncompatibleButton;
@@ -1428,6 +1447,7 @@
         private System.Windows.Forms.ContextMenuStrip ModListHeaderContextMenuStrip;
         private System.Windows.Forms.ToolStripMenuItem reinstallToolStripMenuItem;
         private System.Windows.Forms.ToolStripMenuItem downloadContentsToolStripMenuItem;
+        private System.Windows.Forms.ToolStripMenuItem purgeContentsToolStripMenuItem;
         private CKAN.MainModInfo ModInfoTabControl;
         private System.Windows.Forms.ToolStripStatusLabel StatusLabel;
         private System.Windows.Forms.ToolStripStatusLabel StatusInstanceLabel;

--- a/GUI/Main.cs
+++ b/GUI/Main.cs
@@ -817,9 +817,14 @@ namespace CKAN
             Filter(GUIModFilter.Replaceable);
         }
 
-        private void cachedToolStripMenuItem_Click(object sender, EventArgs e)
+        private void FilterCachedButton_Click(object sender, EventArgs e)
         {
             Filter(GUIModFilter.Cached);
+        }
+
+        private void FilterUncachedButton_Click(object sender, EventArgs e)
+        {
+            Filter(GUIModFilter.Uncached);
         }
 
         private void FilterNewButton_Click(object sender, EventArgs e)
@@ -876,6 +881,7 @@ namespace CKAN
                 case GUIModFilter.InstalledUpdateAvailable: FilterToolButton.Text = Properties.Resources.MainFilterUpgradeable;  break;
                 case GUIModFilter.Replaceable:              FilterToolButton.Text = Properties.Resources.MainFilterReplaceable;  break;
                 case GUIModFilter.Cached:                   FilterToolButton.Text = Properties.Resources.MainFilterCached;       break;
+                case GUIModFilter.Uncached:                 FilterToolButton.Text = Properties.Resources.MainFilterUncached;     break;
                 case GUIModFilter.NewInRepository:          FilterToolButton.Text = Properties.Resources.MainFilterNew;          break;
                 case GUIModFilter.NotInstalled:             ModList.Columns["InstalledVersion"].Visible = false;
                                                             ModList.Columns["InstallDate"].Visible      = false;
@@ -1250,6 +1256,7 @@ namespace CKAN
                 var guiMod = (GUIMod)ModList.Rows[rowIndex].Tag;
 
                 downloadContentsToolStripMenuItem.Enabled = !guiMod.IsCached;
+                purgeContentsToolStripMenuItem.Enabled = guiMod.IsCached;
                 reinstallToolStripMenuItem.Enabled = guiMod.IsInstalled;
             }
         }
@@ -1298,13 +1305,23 @@ namespace CKAN
 
         private void downloadContentsToolStripMenuItem_Click(object sender, EventArgs e)
         {
-            var module = ModInfoTabControl.SelectedModule;
-            if (module == null || !module.IsCKAN)
-                return;
+            ModInfoTabControl.StartDownload(ModInfoTabControl.SelectedModule);
+        }
 
-            Instance.ResetProgress();
-            Instance.ShowWaitDialog(false);
-            ModInfoTabControl.CacheWorker.RunWorkerAsync(module.ToCkanModule());
+        private void purgeContentsToolStripMenuItem_Click(object sender, EventArgs e)
+        {
+            // Purge other versions as well since the user is likely to want that
+            // and has no other way to achieve it
+            if (ModInfoTabControl.SelectedModule != null)
+            {
+                IRegistryQuerier registry = RegistryManager.Instance(CurrentInstance).registry;
+                var allAvail = registry.AllAvailable(ModInfoTabControl.SelectedModule.Identifier);
+                foreach (CkanModule mod in allAvail)
+                {
+                    Manager.Cache.Purge(mod);
+                }
+                ModInfoTabControl.SelectedModule.UpdateIsCached();
+            }
         }
 
         private void Main_Resize(object sender, EventArgs e)

--- a/GUI/Main.cs
+++ b/GUI/Main.cs
@@ -1321,6 +1321,7 @@ namespace CKAN
                     Manager.Cache.Purge(mod);
                 }
                 ModInfoTabControl.SelectedModule.UpdateIsCached();
+                UpdateModContentsTree(ModInfoTabControl.SelectedModule.ToCkanModule(), true);
             }
         }
 

--- a/GUI/Main.resx
+++ b/GUI/Main.resx
@@ -191,7 +191,8 @@
   <data name="FilterInstalledButton.Text" xml:space="preserve"><value>Installed</value></data>
   <data name="FilterInstalledUpdateButton.Text" xml:space="preserve"><value>Installed (update available)</value></data>
   <data name="FilterReplaceableButton.Text" xml:space="preserve"><value>Replaceable</value></data>
-  <data name="cachedToolStripMenuItem.Text" xml:space="preserve"><value>Cached</value></data>
+  <data name="FilterCachedButton.Text" xml:space="preserve"><value>Cached</value></data>
+  <data name="FilterUncachedButton.Text" xml:space="preserve"><value>Uncached</value></data>
   <data name="FilterNewButton.Text" xml:space="preserve"><value>Newly compatible</value></data>
   <data name="FilterNotInstalledButton.Text" xml:space="preserve"><value>Not installed</value></data>
   <data name="FilterIncompatibleButton.Text" xml:space="preserve"><value>Incompatible</value></data>
@@ -213,6 +214,7 @@
   <data name="Description.HeaderText" xml:space="preserve"><value>Description</value></data>
   <data name="reinstallToolStripMenuItem.Text" xml:space="preserve"><value>Reinstall</value></data>
   <data name="downloadContentsToolStripMenuItem.Text" xml:space="preserve"><value>Download Contents</value></data>
+  <data name="purgeContentsToolStripMenuItem.Text" xml:space="preserve"><value>Purge Contents</value></data>
   <data name="ManageModsTabPage.Text" xml:space="preserve"><value>Manage mods</value></data>
   <data name="FilterByAuthorLabel.Text" xml:space="preserve"><value>Filter by author name:</value></data>
   <data name="FilterByAuthorTextBox.Location" type="System.Drawing.Point, System.Drawing"><value>543, 74</value></data>

--- a/GUI/MainModInfo.cs
+++ b/GUI/MainModInfo.cs
@@ -6,6 +6,7 @@ using System.Drawing;
 using System.Windows.Forms;
 using System.IO;
 using System.Diagnostics;
+using System.Threading.Tasks;
 using CKAN.Versioning;
 
 namespace CKAN
@@ -21,16 +22,22 @@ namespace CKAN
 
     public partial class MainModInfo : UserControl
     {
-        private BackgroundWorker m_CacheWorker;
-        private GUIMod _selectedModule;
+        private BackgroundWorker          cacheWorker;
+        private GUIMod                    selectedModule;
+        private NetAsyncModulesDownloader downloader;
+        private CkanModule                currentModContentsModule;
 
         public MainModInfo()
         {
             InitializeComponent();
 
-            m_CacheWorker = new BackgroundWorker { WorkerReportsProgress = true, WorkerSupportsCancellation = true };
-            m_CacheWorker.RunWorkerCompleted += PostModCaching;
-            m_CacheWorker.DoWork += CacheMod;
+            cacheWorker = new BackgroundWorker()
+            {
+                WorkerReportsProgress      = true,
+                WorkerSupportsCancellation = true,
+            };
+            cacheWorker.RunWorkerCompleted += PostModCaching;
+            cacheWorker.DoWork += CacheMod;
 
             DependsGraphTree.BeforeExpand += BeforeExpand;
         }
@@ -39,7 +46,7 @@ namespace CKAN
         {
             set
             {
-                this._selectedModule = value;
+                this.selectedModule = value;
                 if (value == null)
                 {
                     ModInfoTabControl.Enabled = false;
@@ -58,7 +65,7 @@ namespace CKAN
             }
             get
             {
-                return _selectedModule;
+                return selectedModule;
             }
         }
 
@@ -90,18 +97,6 @@ namespace CKAN
             }
         }
 
-        public BackgroundWorker CacheWorker
-        {
-            get
-            {
-                return m_CacheWorker;
-            }
-            set
-            {
-                m_CacheWorker = value;
-            }
-        }
-
         private void DependsGraphTree_NodeMouseDoubleClick(object sender, TreeNodeMouseClickEventArgs e)
         {
             Main.Instance.ResetFilterAndSelectModOnList(e.Node.Name);
@@ -114,14 +109,9 @@ namespace CKAN
 
         private void ContentsDownloadButton_Click(object sender, EventArgs e)
         {
-            var module = SelectedModule;
-            if (module == null || !module.IsCKAN) return;
-
-            Main.Instance.ResetProgress();
-            Main.Instance.ShowWaitDialog(false);
-            m_CacheWorker.RunWorkerAsync(module.ToCkanModule());
+            StartDownload(SelectedModule);
         }
-
+        
         private void ContentsOpenButton_Click(object sender, EventArgs e)
         {
             Process.Start(manager.Cache.GetCachedFilename(SelectedModule.ToModule()));
@@ -407,8 +397,6 @@ namespace CKAN
             Util.Invoke(ContentsPreviewTree, () => _UpdateModContentsTree(force));
         }
 
-        private CkanModule current_mod_contents_module;
-
         private void _UpdateModContentsTree(bool force = false)
         {
             GUIMod guiMod = SelectedModule;
@@ -417,13 +405,13 @@ namespace CKAN
                 return;
             }
             CkanModule module = guiMod.ToCkanModule();
-            if (Equals(module, current_mod_contents_module) && !force)
+            if (Equals(module, currentModContentsModule) && !force)
             {
                 return;
             }
             else
             {
-                current_mod_contents_module = module;
+                currentModContentsModule = module;
             }
             if (!guiMod.IsCached)
             {
@@ -457,29 +445,52 @@ namespace CKAN
             ContentsPreviewTree.Nodes[0].ExpandAll();
         }
 
+        public void StartDownload(GUIMod module)
+        {
+            if (module == null || !module.IsCKAN)
+                return;
+
+            Main.Instance.ShowWaitDialog(false);
+            if (cacheWorker.IsBusy)
+            {
+                Task.Factory.StartNew(() => 
+                {
+                    // Just pass to the existing worker
+                    downloader.DownloadModules(new List<CkanModule> { module.ToCkanModule() });
+                    module.UpdateIsCached();
+                });
+            }
+            else
+            {
+                // Start up a new worker
+                downloader = new NetAsyncModulesDownloader(Main.Instance.currentUser, Main.Instance.Manager.Cache);
+                cacheWorker.RunWorkerAsync(module);
+            }
+        }
+
+        // cacheWorker.DoWork
         private void CacheMod(object sender, DoWorkEventArgs e)
         {
             Main.Instance.ResetProgress();
             Main.Instance.ClearLog();
 
-            NetAsyncModulesDownloader downloader = new NetAsyncModulesDownloader(Main.Instance.currentUser, Main.Instance.Manager.Cache);
-
-            downloader.DownloadModules(new List<CkanModule> { (CkanModule)e.Argument });
+            GUIMod gm = e.Argument as GUIMod;
+            downloader.DownloadModules(new List<CkanModule> { gm.ToCkanModule() });
             e.Result = e.Argument;
         }
 
+        // cacheWorker.RunWorkerCompleted
         public void PostModCaching(object sender, RunWorkerCompletedEventArgs e)
         {
-            Util.Invoke(this, () => _PostModCaching((CkanModule)e.Result));
+            Util.Invoke(this, () => _PostModCaching((GUIMod)e.Result));
         }
 
-        private void _PostModCaching(CkanModule module)
+        private void _PostModCaching(GUIMod module)
         {
+            module.UpdateIsCached();
             Main.Instance.HideWaitDialog(true);
-
-            SelectedModule?.UpdateIsCached();
-            UpdateModContentsTree(module, true);
-            Main.Instance.RecreateDialogs();
+            // User might have selected another row. Show current in tree.
+            UpdateModContentsTree(SelectedModule.ToCkanModule(), true);
         }
 
         /// <summary>

--- a/GUI/MainModList.cs
+++ b/GUI/MainModList.cs
@@ -23,7 +23,8 @@ namespace CKAN
         Incompatible             = 5,
         All                      = 6,
         Cached                   = 7,
-        Replaceable              = 8
+        Replaceable              = 8,
+        Uncached                 = 9,
     }
 
     public partial class Main
@@ -244,26 +245,27 @@ namespace CKAN
             var has_any_installed    = gui_mods.Any(mod => mod.IsInstalled);
             var has_any_replacements = gui_mods.Any(mod => mod.IsInstalled && mod.HasReplacement);
 
-            //TODO Consider using smart enumeration pattern so stuff like this is easier
             Util.Invoke(menuStrip2, () =>
             {
-                FilterToolButton.DropDownItems[0].Text = String.Format(Properties.Resources.MainModListCompatible,
+                FilterCompatibleButton.Text = String.Format(Properties.Resources.MainModListCompatible,
                     mainModList.CountModsByFilter(GUIModFilter.Compatible));
-                FilterToolButton.DropDownItems[1].Text = String.Format(Properties.Resources.MainModListInstalled,
+                FilterInstalledButton.Text = String.Format(Properties.Resources.MainModListInstalled,
                     mainModList.CountModsByFilter(GUIModFilter.Installed));
-                FilterToolButton.DropDownItems[2].Text = String.Format(Properties.Resources.MainModListUpgradeable,
+                FilterInstalledUpdateButton.Text = String.Format(Properties.Resources.MainModListUpgradeable,
                     mainModList.CountModsByFilter(GUIModFilter.InstalledUpdateAvailable));
-                FilterToolButton.DropDownItems[3].Text = String.Format(Properties.Resources.MainModListReplaceable,
+                FilterReplaceableButton.Text = String.Format(Properties.Resources.MainModListReplaceable,
                     mainModList.CountModsByFilter(GUIModFilter.Replaceable));
-                FilterToolButton.DropDownItems[4].Text = String.Format(Properties.Resources.MainModListCached,
+                FilterCachedButton.Text = String.Format(Properties.Resources.MainModListCached,
                     mainModList.CountModsByFilter(GUIModFilter.Cached));
-                FilterToolButton.DropDownItems[5].Text = String.Format(Properties.Resources.MainModListNewlyCompatible,
+                FilterUncachedButton.Text = String.Format(Properties.Resources.MainModListUncached,
+                    mainModList.CountModsByFilter(GUIModFilter.Uncached));
+                FilterNewButton.Text = String.Format(Properties.Resources.MainModListNewlyCompatible,
                     mainModList.CountModsByFilter(GUIModFilter.NewInRepository));
-                FilterToolButton.DropDownItems[6].Text = String.Format(Properties.Resources.MainModListNotInstalled,
+                FilterNotInstalledButton.Text = String.Format(Properties.Resources.MainModListNotInstalled,
                     mainModList.CountModsByFilter(GUIModFilter.NotInstalled));
-                FilterToolButton.DropDownItems[7].Text = String.Format(Properties.Resources.MainModListIncompatible,
+                FilterIncompatibleButton.Text = String.Format(Properties.Resources.MainModListIncompatible,
                     mainModList.CountModsByFilter(GUIModFilter.Incompatible));
-                FilterToolButton.DropDownItems[8].Text = String.Format(Properties.Resources.MainModListAll,
+                FilterAllButton.Text = String.Format(Properties.Resources.MainModListAll,
                     mainModList.CountModsByFilter(GUIModFilter.All));
 
                 UpdateAllToolButton.Enabled = has_any_updates;
@@ -999,6 +1001,7 @@ namespace CKAN
                 case GUIModFilter.Installed:                return m.IsInstalled;
                 case GUIModFilter.InstalledUpdateAvailable: return m.IsInstalled && m.HasUpdate;
                 case GUIModFilter.Cached:                   return m.IsCached;
+                case GUIModFilter.Uncached:                 return !m.IsCached;
                 case GUIModFilter.NewInRepository:          return m.IsNew;
                 case GUIModFilter.NotInstalled:             return !m.IsInstalled;
                 case GUIModFilter.Incompatible:             return m.IsIncompatible;

--- a/GUI/Properties/Resources.Designer.cs
+++ b/GUI/Properties/Resources.Designer.cs
@@ -376,6 +376,9 @@ namespace CKAN.Properties {
         internal static string MainFilterCached {
             get { return (string)(ResourceManager.GetObject("MainFilterCached", resourceCulture)); }
         }
+        internal static string MainFilterUncached {
+            get { return (string)(ResourceManager.GetObject("MainFilterUncached", resourceCulture)); }
+        }
         internal static string MainFilterNew {
             get { return (string)(ResourceManager.GetObject("MainFilterNew", resourceCulture)); }
         }
@@ -585,6 +588,9 @@ namespace CKAN.Properties {
         }
         internal static string MainModListCached {
             get { return (string)(ResourceManager.GetObject("MainModListCached", resourceCulture)); }
+        }
+        internal static string MainModListUncached {
+            get { return (string)(ResourceManager.GetObject("MainModListUncached", resourceCulture)); }
         }
         internal static string MainModListNewlyCompatible {
             get { return (string)(ResourceManager.GetObject("MainModListNewlyCompatible", resourceCulture)); }

--- a/GUI/Properties/Resources.de-DE.resx
+++ b/GUI/Properties/Resources.de-DE.resx
@@ -158,6 +158,7 @@
   <data name="MainFilterUpgradeable" xml:space="preserve"><value>Filter (Aktualisierbar)</value></data>
   <data name="MainFilterReplaceable" xml:space="preserve"><value>Filter (Ersetzbar)</value></data>
   <data name="MainFilterCached" xml:space="preserve"><value>Filter (Im Cache)</value></data>
+  <data name="MainFilterUncached" xml:space="preserve"><value>Filter (Nicht im Cache)</value></data>
   <data name="MainFilterNew" xml:space="preserve"><value>Filter (Neu)</value></data>
   <data name="MainFilterNotInstalled" xml:space="preserve"><value>Filter (Nicht installiert)</value></data>
   <data name="MainFilterCompatible" xml:space="preserve"><value>Filter (Kompatibel)</value></data>
@@ -261,6 +262,7 @@ Einstellungen jetzt öffnen?</value>
   <data name="MainModListUpgradeable" xml:space="preserve"><value>Aktualisierbar ({0})</value></data>
   <data name="MainModListReplaceable" xml:space="preserve"><value>Ersetzbar ({0})</value></data>
   <data name="MainModListCached" xml:space="preserve"><value>Im Cache ({0})</value></data>
+  <data name="MainModListUncached" xml:space="preserve"><value>Nicht im Cache ({0})</value></data>
   <data name="MainModListNewlyCompatible" xml:space="preserve"><value>Neu kompatibel ({0})</value></data>
   <data name="MainModListNotInstalled" xml:space="preserve"><value>Nicht installiert ({0})</value></data>
   <data name="MainModListIncompatible" xml:space="preserve"><value>Inkompatibel ({0})</value></data>

--- a/GUI/Properties/Resources.resx
+++ b/GUI/Properties/Resources.resx
@@ -180,6 +180,7 @@
   <data name="MainFilterUpgradeable" xml:space="preserve"><value>Filter (Upgradeable)</value></data>
   <data name="MainFilterReplaceable" xml:space="preserve"><value>Filter (Replaceable)</value></data>
   <data name="MainFilterCached" xml:space="preserve"><value>Filter (Cached)</value></data>
+  <data name="MainFilterUncached" xml:space="preserve"><value>Filter (Uncached)</value></data>
   <data name="MainFilterNew" xml:space="preserve"><value>Filter (New)</value></data>
   <data name="MainFilterNotInstalled" xml:space="preserve"><value>Filter (Not installed)</value></data>
   <data name="MainFilterCompatible" xml:space="preserve"><value>Filter (Compatible)</value></data>
@@ -282,6 +283,7 @@ Open settings now?</value></data>
   <data name="MainModListUpgradeable" xml:space="preserve"><value>Upgradeable ({0})</value></data>
   <data name="MainModListReplaceable" xml:space="preserve"><value>Replaceable ({0})</value></data>
   <data name="MainModListCached" xml:space="preserve"><value>Cached ({0})</value></data>
+  <data name="MainModListUncached" xml:space="preserve"><value>Uncached ({0})</value></data>
   <data name="MainModListNewlyCompatible" xml:space="preserve"><value>Newly compatible ({0})</value></data>
   <data name="MainModListNotInstalled" xml:space="preserve"><value>Not installed ({0})</value></data>
   <data name="MainModListIncompatible" xml:space="preserve"><value>Incompatible ({0})</value></data>

